catalogue_graph/src/index_remover.py

```
@@ -0,0 +1,141 @@
+from graph_remover import DELETED_IDS_FOLDER
+from utils.aws import df_from_s3_parquet
+
+# This is part of a safety mechanism. If two sets of IDs differ by more than 5%, an exception will be raised.
+ACCEPTABLE_DIFF_THRESHOLD = 0.05
+
+
+def _get_last_index_remover_run_file_uri(pipeline_date: str | None) -> str:
+    return f"s3://{config.INGESTOR_S3_BUCKET}/ingestor/{pipeline_date or 'dev'}/last_index_remover_run_date.txt"
+
+
+def _get_concepts_index_name(pipeline_date: str | None) -> str:
+    return (
+        "concepts-indexed"
+        if pipeline_date is None
+        else f"concepts-indexed-{pipeline_date}"
+    )
+
+
+def get_current_id_count(pipeline_date: str | None, is_local: bool) -> int:
+    """Return the number of documents currently stored in ES in the concepts index."""
+    es = utils.elasticsearch.get_client(pipeline_date, is_local)
+
+    response = es.count(index=_get_concepts_index_name(pipeline_date))
+    count: int = response.get("count", 0)
+    return count
+
+
+def delete_concepts_from_elasticsearch(
+    deleted_ids: set[str], pipeline_date: str | None, is_local: bool
+) -> None:
+    """Remove documents matching `deleted_ids` from the concepts ES index."""
+    es = utils.elasticsearch.get_client(pipeline_date, is_local)
+    index_name = _get_concepts_index_name(pipeline_date)
+
+    response = es.delete_by_query(
+        index=index_name, body={"query": {"ids": {"values": list(deleted_ids)}}}
+    )
+
+    deleted_count = response["deleted"]
+    print(f"Deleted {deleted_count} documents from the {index_name} index.")
+
+
+def get_ids_to_delete(pipeline_date: str | None) -> set[str]:
+    """Return a list of concept IDs marked for deletion from the ES index."""
+    s3_file_uri = f"s3://{config.INGESTOR_S3_BUCKET}/{DELETED_IDS_FOLDER}/catalogue_concepts__nodes.parquet"
+
+    # Retrieve a log of concept IDs which were deleted from the graph (see `graph_remover.py`).
+    df = df_from_s3_parquet(s3_file_uri)
```

I think this will fail if the file does not exist, is that the behaviour we want here or would it be preferable to succeed but do no work? 

It would be unexpected for this to run without the graph remover having been run first so a failure is probably correct, but in that case perhaps we should have a test for it?

---

Yes, I think we should let the code fail if the file doesn't exist. If that happens, something went wrong.

I added a new test to cover this.